### PR TITLE
prefer base-compat to CPP macros

### PR DESCRIPTION
**Pull request by dredozubov**
basically #403

[Review on Reviewable](https://reviewable.io/reviews/haskell-servant/servant/404)

### CODE REVIEW

File: servant-server/src/Servant/Server/Internal.hs [hunk 1]
```diff
@@ -43,6 +40,7 @@ import           Network.Wai                (Application, Request, Response,
                                              rawQueryString, remoteHost,
                                              requestHeaders, requestMethod,
                                              responseLBS, vault)
+import           Prelude.Compat
```
Comment (sol) [inline servant-server/src/Servant/Server/Internal.hs]: I guess this should be:

```haskell
import           Prelude ()
import           Prelude.Compat
```

File: servant/src/Servant/API/ContentTypes.hs [hunk 2]
```diff
@@ -72,11 +72,8 @@ module Servant.API.ContentTypes
     , canHandleAcceptH
     ) where
 
-#if !MIN_VERSION_base(4,8,0)
-import           Control.Applicative              ((*>), (<*))
-#endif
 import           Control.Arrow                    (left)
-import           Control.Monad
+import           Control.Monad.Compat             hiding (mapM)
```
Comment (sol) [inline servant/src/Servant/API/ContentTypes.hs]: Similarly here, if you use `base-compat`, you want to use `*.Compat` modules consistently to prevent name collisions.

**Comment by arianvp** [comment]
Great idea.  I really don't like all the CPP that we have scattered around the library currently.

File: servant/src/Servant/API/ContentTypes.hs (hunk 2 above)
Comment (dredozubov) [inline servant/src/Servant/API/ContentTypes.hs]: thanks for correcting me, i'll update the PR

**Comment by dredozubov** [comment]
I've updated the PR according to @sol comments.

File: servant/src/Servant/API/ContentTypes.hs (hunk 2 above)
Comment (sol) [inline servant/src/Servant/API/ContentTypes.hs]: I'm still puzzled about the `hiding (mapM)`.  Is this necessary?

File: servant/src/Servant/Utils/Links.hs [hunk 3]
```diff
@@ -105,11 +105,7 @@ import Data.List
 import Data.Proxy ( Proxy(..) )
 import qualified Data.Text as Text
 import qualified Data.ByteString.Char8 as BSC
-#if !MIN_VERSION_base(4,8,0)
-import Data.Monoid ( Monoid(..), (<>) )
-#else
-import Data.Monoid ( (<>) )
-#endif
+import Data.Monoid.Compat ( Monoid(..), (<>) )
```
Comment (sol) [inline servant/src/Servant/Utils/Links.hs]: I think if you import `Prelude.Compat` here, you don't need to import the `Monoid` type class.   That may be a more robust solution.  You still will need `(<>)` though...

**Comment by sol** [comment]
Except for my comments, LGTM.

File: servant/src/Servant/Utils/Links.hs (hunk 3 above)
Comment (dredozubov) [inline servant/src/Servant/Utils/Links.hs]: what's the point then?

**Comment by dredozubov** [comment]
Ehm, weird:

```
% git diff origin/master origin/base-compat servant/src/Servant/API/ContentTypes.hs
diff --git a/servant/src/Servant/API/ContentTypes.hs b/servant/src/Servant/API/ContentTypes.hs
index 61bf1ce..bed009a 100644
--- a/servant/src/Servant/API/ContentTypes.hs
+++ b/servant/src/Servant/API/ContentTypes.hs
@@ -72,11 +72,8 @@ module Servant.API.ContentTypes
     , canHandleAcceptH
     ) where

-#if !MIN_VERSION_base(4,8,0)
-import           Control.Applicative              ((*>), (<*))
-#endif
 import           Control.Arrow                    (left)
-import           Control.Monad
+import           Control.Monad.Compat             hiding (mapM)
 import           Data.Aeson                       (FromJSON(..), ToJSON(..), encode)
 import           Data.Aeson.Parser                (value)
 import           Data.Aeson.Types                 (parseEither)
@@ -88,7 +85,7 @@ import           Data.ByteString.Lazy             (ByteString, fromStrict,
 import qualified Data.ByteString.Lazy             as B
 import qualified Data.ByteString.Lazy.Char8       as BC
 import           Data.Maybe                       (isJust)
-import           Data.Monoid
+import           Data.Monoid.Compat
 import           Data.String.Conversions          (cs)
 import qualified Data.Text                        as TextS
 import qualified Data.Text.Encoding               as TextS
@@ -99,6 +96,8 @@ import           GHC.Generics                     (Generic)
 import qualified Network.HTTP.Media               as M
 import           Network.URI                      (escapeURIString,
                                                    isUnreserved, unEscapeString)
+import           Prelude                          ()
+import           Prelude.Compat

 -- * Provided content types
 data JSON deriving Typeable
```

```
git remote -v | ag origin
origin	git@github.com:haskell-servant/servant.git (fetch)
origin	git@github.com:haskell-servant/servant.git (push)
```

File: servant/src/Servant/API/ContentTypes.hs (hunk 2 above)
Comment (dredozubov) [inline servant/src/Servant/API/ContentTypes.hs]: check checked, i guess it became unnecessary after the recent changes

File: servant/src/Servant/Utils/Links.hs (hunk 3 above)
Comment (soenkehahn) [inline servant/src/Servant/Utils/Links.hs]: IIUC `Monoid` is being exported by newer `Prelude`s (and hence also by `Prelude.Compat`). `(<>)` is not being exported by newer `Prelude`s (and hence also not by `Prelude.Compat`). So what we want here, I guess, is:

``` haskell
import Prelude ()
import Prelude.Compat
import Data.Monoid ((<>))
```
The point is: less CPP.

**Comment by dredozubov** [comment]
@soenkehahn @sol any more suggestions?
@jkarni are you ok with this PR?

File: servant-foreign/servant-foreign.cabal [hunk 4]
```diff
@@ -27,10 +27,10 @@ source-repository head
...
 library
   exposed-modules:     Servant.Foreign, Servant.Foreign.Internal
-  build-depends:       base       == 4.*
-                     , lens       == 4.*
-                     , servant    == 0.5.*
-                     , text       >= 1.2  && < 1.3
+  build-depends:       base >=4.7 && <5
+                     , lens        == 4.*
+                     , servant     == 0.5.*
+                     , text        >= 1.2  && < 1.3
```
Comment (soenkehahn) [inline servant-foreign/servant-foreign.cabal]: Why was this changed? Both the whitespace and the narrower version bounds for base?

File: servant/src/Servant/API/Internal/Test/ComprehensiveAPI.hs [hunk 5]
```diff
@@ -7,7 +7,6 @@
 module Servant.API.Internal.Test.ComprehensiveAPI where
 
 import           Data.Proxy
-
```
Comment (soenkehahn) [inline servant/src/Servant/API/Internal/Test/ComprehensiveAPI.hs]: I guess this line was to separate external imports from internal ones.

**Comment by soenkehahn** [comment]
@dredozubov: Did you make sure to remove `CPP` language extensions where possible?

Apart from my comments LGTM. Thanks for working on this!

**Comment by arianvp** [comment]
The following files still use CPP in this PR:

```
servant-blaze/src/Servant/HTML/Blaze.hs:{-# LANGUAGE CPP                   #-}
servant-cassava/src/Servant/CSV/Cassava.hs:{-# LANGUAGE CPP                   #-}
servant-client/src/Servant/Client.hs:{-# LANGUAGE CPP                  #-}
servant-client/src/Servant/Common/Req.hs:{-# LANGUAGE CPP                 #-}
servant-client/test/Servant/ClientSpec.hs:{-# LANGUAGE CPP                    #-}
servant-client/test/Servant/Common/BaseUrlSpec.hs:{-# LANGUAGE CPP #-}
servant-docs/src/Servant/Docs/Internal.hs:{-# LANGUAGE CPP                    #-}
servant-examples/socket-io-chat/Chat.hs:{-# LANGUAGE CPP #-}
servant-examples/socket-io-chat/socket-io-chat.hs:{-# LANGUAGE CPP               #-}
servant-examples/tutorial/T1.hs:{-# LANGUAGE CPP           #-}
servant-examples/tutorial/T2.hs:{-# LANGUAGE CPP           #-}
servant-foreign/src/Servant/Foreign/Internal.hs:{-# LANGUAGE CPP                   #-}
servant-foreign/test/Servant/ForeignSpec.hs:{-# LANGUAGE CPP                   #-}
servant-js/test/Servant/JSSpec.hs:{-# LANGUAGE CPP                 #-}
servant-lucid/src/Servant/HTML/Lucid.hs:{-# LANGUAGE CPP                   #-}
servant-mock/src/Servant/Mock.hs:{-# LANGUAGE CPP                   #-}
servant-server/src/Servant/Server/Internal/Config.hs:{-# LANGUAGE CPP                        #-}
servant-server/src/Servant/Server/Internal/Enter.hs:{-# LANGUAGE CPP                    #-}
servant-server/src/Servant/Server/Internal.hs:{-# LANGUAGE CPP                        #-}
servant-server/src/Servant/Server/Internal.hs:ct_wildcard = "*" <> "/" <> "*" -- Because CPP
servant-server/src/Servant/Server/Internal/Router.hs:{-# LANGUAGE CPP #-}
servant-server/src/Servant/Server/Internal/RoutingApplication.hs:{-# LANGUAGE CPP                        #-}
servant-server/src/Servant/Utils/StaticFiles.hs:{-# LANGUAGE CPP #-}
servant-server/test/Servant/ServerSpec.hs:{-# LANGUAGE CPP                  #-}
servant/src/Servant/API/Alternative.hs:{-# LANGUAGE CPP                #-}
servant/src/Servant/API/ContentTypes.hs:{-# LANGUAGE CPP                   #-}
servant/src/Servant/API/ResponseHeaders.hs:{-# LANGUAGE CPP                    #-}
servant/src/Servant/Utils/Links.hs:{-# LANGUAGE CPP                    #-}
servant/test/Servant/API/ContentTypesSpec.hs:{-# LANGUAGE CPP 
```

Perhaps we can systematically go through them and see how much of it can be removed.

**Comment by dredozubov** [comment]
@arianvp @soenkehahn completely removing CPP pragmas will require purging ```#include "overlapping-compat.h"``` from these modules, which will break compatibility with older GHC versions

File: servant-foreign/servant-foreign.cabal [hunk 6]
```diff
@@ -27,10 +27,10 @@ source-repository head
...
 library
   exposed-modules:     Servant.Foreign, Servant.Foreign.Internal
-  build-depends:       base       == 4.*
-                     , lens       == 4.*
-                     , servant    == 0.5.*
-                     , text       >= 1.2  && < 1.3
+  build-depends:       base        == 4.*
+                     , lens        == 4.*
+                     , servant     == 0.5.*
+                     , text        >= 1.2  && < 1.3
```
Comment (soenkehahn) [inline servant-foreign/servant-foreign.cabal]: Can you remove this whitespace change, please?

**Comment by soenkehahn** [comment]
@dredozubov: Yes, we definitely shouldn't increase the scope of this PR. (And I believe @arianvp suggested to look through uses of `CPP` separate from this.)

I only meant to ask whether you tried to remove the CPP pragmas to see whether it was possible. If there are other uses of `CPP` in some modules, of course, the pragmas have to stay.

Feel free to merge.

**Comment by jkarni** [comment]
:+1:

**Comment by soenkehahn** [comment]
Awesome. @dredozubov: Thanks again a lot for getting this done.
